pkg/protodep/gomod.go

```diff
@@ -0,0 +1,331 @@
+	"github.com/spf13/afero"
+)
+
+const (
+	ProtoMatchPattern   = "**/*.proto"
+	SoloKitMatchPattern = "**/solo-kit.json"
+)
+
+var (
+	// offer sane defaults for proto vendoring
+	DefaultMatchPatterns = []string{ProtoMatchPattern, SoloKitMatchPattern}
+
+	// matches ext.proto for solo hash gen
+	ExtProtoMatcher = &GoModImport{
+		Package:  "github.com/solo-io/protoc-gen-ext",
+		Patterns: []string{"extproto/*.proto"},
+	}
+
+	// matches validate.proto which is needed by envoy protos
+	ValidateProtoMatcher = &GoModImport{
+		Package:  "github.com/envoyproxy/protoc-gen-validate",
+		Patterns: []string{"validate/*.proto"},
+	}
+
+	// matches all solo-kit protos, useful for any projects using solo-kit
+	SoloKitProtoMatcher = &GoModImport{
+		Package:  "github.com/solo-io/solo-kit",
+		Patterns: []string{"api/**/*.proto", "api/" + SoloKitMatchPattern},
+	}
+
+	// matches gogo.proto, used for gogoproto code gen.
+	GogoProtoMatcher = &GoModImport{
+		Package:  "github.com/gogo/protobuf",
+		Patterns: []string{"gogoproto/*.proto"},
+	}
+
+	// default match options which should be used when creating a solo-kit project
+	DefaultMatchOptions = []*Import{
+		{
+			ImportType: &Import_GoMod{
+				GoMod: ExtProtoMatcher,
+			},
+		},
+		{
+			ImportType: &Import_GoMod{
+				GoMod: ValidateProtoMatcher,
+			},
+		},
+		{
+			ImportType: &Import_GoMod{
+				GoMod: SoloKitProtoMatcher,
+			},
+		},
+		{
+			ImportType: &Import_GoMod{
+				GoMod: GogoProtoMatcher,
+			},
+		},
+	}
+)
+
+type goModOptions struct {
+	MatchOptions  []*GoModImport
+	LocalMatchers []string
+}
+
+// struct which represents a go module package in the module package list
+type Module struct {
+	ImportPath     string
+	SourcePath     string
+	Version        string
+	SourceVersion  string
+	Dir            string   // full path, $GOPATH/pkg/mod/
+	VendorList     []string // files to vendor
+	currentPackage bool
+}
+
+func NewGoModFactory(cwd string) (*goModFactory, error) {
+	if !filepath.IsAbs(cwd) {
+		absoluteDir, err := filepath.Abs(cwd)
+		if err != nil {
+			return nil, err
+		}
+		cwd = absoluteDir
+	}
+	fs := afero.NewOsFs()
+	return &goModFactory{
+		WorkingDirectory: cwd,
+		fs:               fs,
+		cp:               NewCopier(fs),
+	}, nil
+}
+
+type goModFactory struct {
+	WorkingDirectory string
+	packageName      bool
+	fs               afero.Fs
+	cp               FileCopier
+}
+
+func (m *goModFactory) Ensure(ctx context.Context, opts *Config) error {
+	var packages []*GoModImport
+	for _, cfg := range opts.Imports {
+		if cfg.GetGoMod() != nil {
+			packages = append(packages, cfg.GetGoMod())
+		}
+	}
+	mods, err := m.gather(goModOptions{
+		MatchOptions:  packages,
+		LocalMatchers: opts.GetLocal().GetPatterns(),
+	})
+	if err != nil {
+		return err
+	}
+
+	err = m.copy(mods)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// gather up all packages for a given go module
+// currently this function uses the cmd `go list -m all` to figure out the list of dep
```

Comment:
I appreciate this note but it's not clear where `go list -m all` is called - might be worth indicating where this is happening